sycl/include/CL/sycl/ONEAPI/group_algorithm.hpp

```diff
@@ -190,6 +190,28 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
         !cl::sycl::detail::is_native_op<T, BinaryOperation>::value,
     T>;
 
+template <typename T>
+struct is_native_function_object_impl : std::false_type {};
+
+template <template <typename> class BinaryOperation, typename T>
+struct is_native_function_object_impl<BinaryOperation<T>>
+    : std::integral_constant<
+          bool, (sycl::detail::is_scalar_arithmetic<T>::value ||
+                 sycl::detail::is_vector_arithmetic<T>::value ||
+                 std::is_same<T, void>::value) &&
+                    sycl::detail::is_native_op<T, BinaryOperation<T>>::value> {
+};
```

Comment:
I mentioned it as a minor in the other thread. I don't think there is any value in checking anything about `T`. Nor does the spec say anything about it. 

```suggestion
    : sycl::detail::is_native_op<T, BinaryOperation<T>> {};
```

Reply:
This is much harder than I anticipated.  You're right that the specification doesn't say anything about checking `T`, but maybe it should.

You'd almost convinced me (and I started implementing your suggestion), but then I realized that if SYCL itself doesn't provide a trait that checks the combination of `T` and `BinaryOperation` it's much harder for the user to do it themselves.

Consider the below:
```c++
is_native_function_object_v<BinaryOperation> && /* T is a SYCL fundamental type */
```

With your proposed patch, this will return true if `BinaryOperation` is `plus<T2>` but the inputs are `T`.  So we'd only be providing part of the logic ("is this function object one that SYCL knows something about?").  Users would have to write individual checks for each algorithm, right?